### ISSUE TITLE
support of not_inside with random declared list

### ISSUE THREAD

**Issue by ShraddhaDevaiya** [fvutils/pyvsc#31]
Hi @mballance , I am trying to use not_inside feature with list declared random variable. Following is the code for it : 

```
import vsc


@vsc.randobj
class my_s(object):
    def __init__(self):
        self.temp = vsc.randsz_list_t(vsc.uint8_t(8))
        self.a = [1,2,3,5,6,7,8]

    @vsc.constraint
    def ls(self):
        with vsc.foreach(self.temp, idx = True) as i:
            self.temp[i].not_inside(vsc.rangelist(self.a))

io = my_s()
io.randomize()
print(io.temp)
```

And for this, it is giving an error like following:
```
Exception while processing constraint: 'expr' object has no attribute 'not_inside'
Traceback (most recent call last):
  File "unique_list.py", line 16, in <module>
    io = my_s()
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/rand_obj.py", line 61, in __init__
    self.build_field_model(None)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/rand_obj.py", line 134, in build_field_model
    raise e
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/rand_obj.py", line 131, in build_field_model
    fo.c(self)
  File "unique_list.py", line 13, in ls
    self.temp[i].not_inside(vsc.rangelist(self.a))
AttributeError: 'expr' object has no attribute 'not_inside'
```

With this error, it seems that it is not supported for **rand_list_t** and **randsz_list_t** declared variables.
Can you please take a look at this?

**Comment by mballance** [comment]
Hi @ShraddhaDevaiya,
    Good catch on this one. Because of how array constraints are implemented, a reference to an array element inside a foreach constraint is captured as an expression initially. I've corrected this in the 0.1.3 PyPi release.
    Please note that you'll need to add a size-bounding constraint on the random-size list. For example:
```
self.temp.size <= 10
```

Best Regards,
Matthew

**Comment by ShraddhaDevaiya** [comment]
Hi @mballance , thanks for this quick fix. I have checked it, it is not giving compilation error for this , but sometimes randomization is giving wrong values. Like with previously given code, it is giving an output like following : 
```
[65, 52, 250, 142, 21, 7, 234, 6, 16, 140]
```
In this, according to constraint 6 and 7 should not come.

And one more thing if I declare list with this : ```self.a = vsc.list_t(vsc.uint8_t(),7)```

this is giving an error like following : 
```
Traceback (most recent call last):
  File "unique_list.py", line 20, in <module>
    io.randomize()
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/rand_obj.py", line 107, in randomize
    Randomizer.do_randomize([model])
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/randomizer.py", line 830, in do_randomize
    bounds_v.process(field_model_l, constraint_l)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/visitors/variable_bound_visitor.py", line 69, in process
    v.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/field_composite_model.py", line 143, in accept
    v.visit_composite_field(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/model_visitor.py", line 64, in visit_composite_field
    c.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/constraint_block_model.py", line 42, in accept
    v.visit_constraint_block(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/model_visitor.py", line 100, in visit_constraint_block
    self.visit_constraint_scope(c)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/model_visitor.py", line 139, in visit_constraint_scope
    cc.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/constraint_override_model.py", line 22, in accept
    v.visit_constraint_override(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/model_visitor.py", line 135, in visit_constraint_override
    c.new_constraint.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/constraint_inline_scope_model.py", line 15, in accept
    v.visit_constraint_inline_scope(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/visitors/variable_bound_visitor.py", line 91, in visit_constraint_inline_scope
    cc.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/constraint_expr_model.py", line 37, in accept
    visitor.visit_constraint_expr(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/model_visitor.py", line 107, in visit_constraint_expr
    c.e.accept(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/model/expr_fieldref_model.py", line 47, in accept
    visitor.visit_expr_fieldref(self)
  File "/home/shraddha/.local/lib/python3.6/site-packages/vsc/visitors/variable_bound_visitor.py", line 259, in visit_expr_fieldref
    raise Exception("Field " + e.fm.name + " not in map")
Exception: Field <unknown> not in map
```
Can you please check this?

**Comment by mballance** [comment]
Hi @ShraddhaDevaiya,
    I've started looking into the first issue, and can see the same thing. My apologies, I should have looked more closely at the test results before considering this resolved. I'll need to spend more time investigating, but wanted to let you know first that we're seeing the same issue.

Best Regards,
Matthew

**Comment by mballance** [comment]
Hi @ShraddhaDevaiya ,
    I believe I've resolved this issue with the 0.1.5 release (coming within the next hour or two). I was able to reproduce your first issue, but wasn't able to follow exactly the situation in your second issue. Please provide a testcase if this continues to be an issue.

Best Regards,
Matthew

**Comment by ShraddhaDevaiya** [comment]
Hi @mballance , I have checked this . It is working now. But one thing that I observed for **enum type** values in randomization. Following is the code for this : 
```
import vsc
from enum import Enum,auto

class my_e(Enum):
    A = 0
    B = auto()
    C = auto()
    D = auto()
    E = auto()
    F = auto()


@vsc.randobj
class my_s(object):
    def __init__(self):
        self.temp = vsc.rand_list_t(vsc.enum_t(my_e),10)
        self.a = vsc.list_t(vsc.enum_t(my_e))
        self.a = [my_e.A,my_e.C]


    @vsc.constraint
    def ls(self):
        with vsc.foreach(self.temp, idx = True) as i:
            self.temp[i].not_inside(self.a)

io = my_s()
io.randomize()
print(io.temp)
```
for this it is giving values like shown below : 
```
[my_e.B, my_e.B, my_e.B, my_e.B, my_e.B, my_e.B, my_e.D, my_e.B, my_e.B, my_e.B]
```
means it is not giving other values, it is not shuffled well with available values in **my_e class**. if we get more different values, it will be more efficient. Can you please check this if possible?
